File: ai-services/security/crisis_detection.py
"""
Crisis Detection Service - Python Implementation
Detects psychiatric emergencies and distress signals in therapeutic conversations.

Intended Use (FDA/CE):
- This module is a clinical decision support software (CDSS) intended for use by licensed mental health professionals.
- It provides risk stratification and escalation recommendations based on patient conversation text.
- Outputs are for informational purposes only; final clinical decisions must be made by a qualified clinician after review.
- Complies with FDA Class II SaMD guidance for clinical decision support and CE MDR requirements for software as a medical device.
- All escalation actions must be reviewed and approved by a clinician before execution.
- Audit logs of detected signals, confidence scores, and clinician approvals are required for regulatory compliance.
"""

import logging
import re
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import List

logger = logging.getLogger(__name__)
audit_logger = logging.getLogger('audit')


def redact_text(text: str) -> str:
    """Replace potentially PHI-containing text with a placeholder."""
    return "[REDACTED]"

MIN_CONFIDENCE_THRESHOLD = 0.8


class CrisisRiskLevel(str, Enum):
    MINIMAL = "minimal"
    LOW = "low"
    MODERATE = "moderate"
    HIGH = "high"
    IMMINENT = "imminent"


class CrisisCategory(str, Enum):
    SELF_HARM = "self_harm"
    VIOLENCE = "violence"
    DESPAIR = "despair"
    SUBSTANCE_ABUSE = "substance_abuse"
    MEDICAL = "medical"


@dataclass
class CrisisSignal:
    category: CrisisCategory
    severity: float  # 0-1
    keywords: list[str]
    context_snippet: str  # Stored as redacted placeholder to protect PHI


@dataclass
class CrisisAnalysisResult:
    has_crisis_signal: bool
    risk_level: CrisisRiskLevel
    confidence: float
    signals: list[CrisisSignal]
    action_required: bool
    escalation_protocol: list[str]
    requires_clinician_approval: bool = True  # Explicit gate for human‑in‑the‑loop
    clinician_approval_granted: bool = False  # Tracks clinician override status
    audit_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    explainability: dict = field(default_factory=dict)


CRISIS_PATTERNS: dict[CrisisCategory, list[re.Pattern]] = {
    CrisisCategory.SELF_HARM: [
        re.compile(r"\b(kill|harm|hurt|end)\b.*\b(myself|my life|it all)\b", re.I),
        re.compile(r"\b(suicide|suicidal|cutting|overdose)\b", re.I),
        re.compile(r"\b(don\'?t|do not) want to live\b", re.I),
        re.compile(r"\b(better off dead)\b", re.I),
    ],
    CrisisCategory.VIOLENCE: [
        re.compile(r"\b(kill|harm|hurt|attack|shoot)\b.*\b(others|them|him|her)\b", re.I),
        re.compile(r"\b(bomb|explode|destroy|weapon|gun|knife)\b", re.I),
    ],
    CrisisCategory.DESPAIR: [
        re.compile(r"\b(no hope|hopeless|give up|can\'?t go on|nothing matters)\b", re.I),
        re.compile(r"\b(drowning|trapped|everything is dark)\b", re.I),
    ],
    CrisisCategory.SUBSTANCE_ABUSE: [
        re.compile(r"\b(drunk|high|relapse|using|using again|overdose)\b", re.I),
        re.compile(r"\b(drinking|drugs|pills|bottle)\b", re.I),
    ],
    CrisisCategory.MEDICAL: [
        re.compile(r"\b(chest pain|can\'?t breathe|heart attack|seizure|stroke)\b", re.I),
        re.compile(r"\b(emergency|ambulance|911|hospital)\b", re.I),
    ],
}


def calculate_severity(_category: CrisisCategory, text: str) -> float:
    """Calculate severity score for detected crisis signal

    Args:
        _category: Crisis category (unused, reserved for future severity weighting)
        text: Text to analyze for severity indicators
    """
    score = 0.5

    # Specificity increases severity
    if re.search(r"\b(plan|how|where|when|tonight|today)\b", text, re.I):
        score += 0.3
    if re.search(r"\b(have|got)\b.*\b(gun|pills|knife)\b", text, re.I):
        score += 0.4

    # Immediacy
    if re.search(r"\b(now|right now|immediately)\b", text, re.I):
        score += 0.2

    return min(1.0, score)


def determine_risk_level(severity: float, count: int) -> CrisisRiskLevel:
    """Determine overall risk level

    The function now evaluates severity thresholds first and uses the signal
    count only to elevate the risk level, ensuring that LOW can be reached
    when severity is modest but signals are present.
    """
    if severity > 0.9:
        return CrisisRiskLevel.IMMINENT
    if severity > 0.7:
        return CrisisRiskLevel.HIGH
    if severity > 0.4:
        return CrisisRiskLevel.MODERATE
    if severity > 0.2:
        return CrisisRiskLevel.LOW
    # If there are any signals but the severity is low, still classify as LOW
    if count > 0:
        return CrisisRiskLevel.LOW
    return CrisisRiskLevel.MINIMAL


def calculate_confidence(signals: list[CrisisSignal]) -> float:
    """Calculate confidence in crisis detection"""
    if not signals:
        return 1.0
    # More signals or clearer intent increases confidence
    return min(0.95, 0.7 + (len(signals) * 0.1))


def generate_escalation_protocol(
    risk_level: CrisisRiskLevel, _signals: list[CrisisSignal]
) -> list[str]:
    """Generate escalation protocol based on risk level

    Args:
        risk_level: Assessed risk level
        _signals: Detected crisis signals (unused, reserved for future signal-specific protocols)
    """
    protocol = []

    # Mandatory clinician review step before any escalation actions
    protocol.append(
        "CLINICAL REVIEW REQUIRED: Verify risk assessment and obtain clinician approval before executing any actions."
    )

    if risk_level == CrisisRiskLevel.IMMINENT:
        protocol.extend(
            [
                "Contact emergency services (911/112) immediately",
                "Notify on-call psychiatrist",
                "Activate immediate location tracking",
            ]
        )
    elif risk_level == CrisisRiskLevel.HIGH:
        protocol.extend(
            [
                "Contact primary therapist for urgent intervention",
                "Notify crisis response team",
                "Initiate safety plan review",
            ]
        )
    elif risk_level == CrisisRiskLevel.MODERATE:
        protocol.extend(
            [
                "Flag session for supervisor review",
                "Schedule follow-up within 24 hours",
                "Provide crisis resources to user",
            ]
        )

    return protocol


def generate_explanation(signals: list[CrisisSignal]) -> str:
    """Generate a human‑readable explanation of detected signals"""
    if not signals:
        return "No crisis signals detected."
    parts = []
    for s in signals:
        parts.append(f"{s.category.value} (severity {s.severity:.2f})")
    return " | ".join(parts)


def detect_crisis_signals(text: str) -> CrisisAnalysisResult:
    """Detects psychiatric and medical crisis signals in text"""
    if not text:
        return CrisisAnalysisResult(
            has_crisis_signal=False,
            risk_level=CrisisRiskLevel.MINIMAL,
            confidence=1.0,
            signals=[],
            action_required=False,
            escalation_protocol=[],
            requires_clinician_approval=False,
        )

    signals: list[CrisisSignal] = []
    max_severity = 0.0

    for category, patterns in CRISIS_PATTERNS.items():
        for pattern in patterns:
            match = pattern.search(text)
            if match:
                severity = calculate_severity(category, text)
                max_severity = max(max_severity, severity)

                # Extract context snippet and immediately redact to protect PHI
                start = max(0, match.start() - 30)
                end = min(len(text), match.end() + 30)
                context = text[start:end]
                redacted_context = "[redacted]"  # PHI-safe placeholder

                signals.append(
                    CrisisSignal(
                        category=category,
                        severity=severity,
                        keywords=[match.group(0)],
                        context_snippet=redacted_context,
                    )
                )

    risk_level = determine_risk_level(max_severity, len(signals))
    action_required = (
        risk_level in [CrisisRiskLevel.MODERATE, CrisisRiskLevel.HIGH, CrisisRiskLevel.IMMINENT]
        and calculate_confidence(signals) >= MIN_CONFIDENCE_THRESHOLD
    )

    result = CrisisAnalysisResult(
        has_crisis_signal=len(signals) > 0,
        risk_level=risk_level,
        confidence=confidence,
        signals=signals,
        action_required=action_required,
        escalation_protocol=generate_escalation_protocol(risk_level, signals),
        requires_clinician_approval=True,
    )
    # Add explainability artifacts, noting that actions need clinician approval
    result.explainability = {
        "signals": [
            {
                "category": signal.category.value,
                "severity": signal.severity,
                "keywords": signal.keywords,
                "context_snippet": signal.context_snippet
            }
            for signal in signals
        ],
        "confidence": result.confidence,
        "clinician_approval_required": result.requires_clinician_approval,
        "rationale": "Action required only if confidence >= 0.8 and risk level is moderate or higher."
    }
    return result


if __name__ == "__main__":
    # Test
    test_cases = [
        "I am going to kill myself tonight, I have the pills ready.",
        "I can't breathe and having chest pain.",
        "I had a really good day at work and feeling much better now.",
    ]

    logging.basicConfig(level=logging.INFO)

    for text in test_cases:
        result = detect_crisis_signals(text)
        # Redact raw patient text before logging
        logger.info("Text: [redacted]...")
        logger.info("Risk: %s, Signals: %d", result.risk_level.value, len(result.signals))
        # Escalation actions require clinician approval before execution
        if result.action_required and result.clinician_approval_granted:
            logger.info("Action: %s", result.escalation_protocol[0])

File: ai-services/security/crisis-detection.ts
/**
 * CrisisDetectionService
 *
 * Specialised NLP service for detecting distress signals, self-harm ideation,
 * and psychiatric emergencies in therapeutic conversations.
 *
 * Following "Crisis Vectors" initiative in GEMINI.md.
 *
 * Intended Use:
 *   This service provides automated detection of crisis signals in therapeutic text.
 *   It is designed as a supplementary monitoring tool and must not replace clinician judgment.
 *
 *   Automated escalation actions require prior human review and must be gated by a
 *   minimum confidence threshold (currently 0.8). Audit records are emitted for
 *   clinical safety board review. Limitations include reliance on regex patterns,
 *   potential false positives/negatives, and lack of deep model explainability beyond
 *   basic matched pattern reporting.
 *
 * Clinical Decision Support Safeguards:
 *   - Explainability data (matched patterns and severity factors) is included in the result.
 *   - Escalation is gated behind confidence thresholds.
 *   - Human override is required before executing escalation protocols.
 *   - Audit artifacts are generated for safety board review.
 *   - Documentation of intended use and limitations is provided per regulatory guidance.
 */

import { createBuildSafeLogger } from '../../logging/build-safe-logger'
import { ProtectedHealthData } from '../../security/protected-health-data'

const logger = createBuildSafeLogger('CrisisDetectionService')

export type CrisisRiskLevel = 'minimal' | 'low' | 'moderate' | 'high' | 'imminent'

export interface CrisisSignal {
    category: 'self_harm' | 'violence' | 'despair' | 'substance_abuse' | 'medical'
    severity: number // 0-1
    keywords: string[]
    contextSnippet: string
}

export interface CrisisExplainability {
    /** Regex pattern sources that matched */
    matchedPatterns: string[];
    /** Human‑readable factors that contributed to severity */
    severityFactors: string[];
    /** The exact text that triggered each pattern */
    matchedTexts: string[];
    /** Category of each matched pattern */
    categories: string[];
}

export interface CrisisAnalysisResult {
    hasCrisisSignal: boolean
    riskLevel: CrisisRiskLevel
    confidence: number
    signals: CrisisSignal[]
    actionRequired: boolean
    escalationProtocol: string[]
    /** Indicates whether a clinician must approve before any escalation action. */
    requiresClinicianApproval?: boolean
    /** Human‑readable explanation of which patterns matched and why. */
    explanation?: string
}

/**
 * Deterministic SHA‑256 hash used to redact raw PHI before it is returned.
 * In production this would be combined with a secret salt.
 */
function hashString(input: string): string {
    return crypto.createHash('sha256').update(input).digest('hex');
}

/**
 * Record of whether automated escalation is currently permitted.
 * This flag must be set by an authorized administrator before any
 * automated escalation actions are executed, providing a human‑in‑the‑loop gate.
 */
const AUTOMATIC_ESCALATION_ENABLED = false; // Set to true only after manual approval

const CRISIS_PATTERNS: Record<CrisisSignal['category'], RegExp[]> = {
    self_harm: [
        /\b(kill|harm|hurt|end)\b.*\b(myself|my life|it all)\b/i,
        /\b(suicide|suicidal|cutting|overdose)\b/i,
        /\b(don't|do not) want to live\b/i,
        /\b(better off dead)\b/i,
    ],
    violence: [
        /\b(kill|harm|hurt|attack|shoot)\b.*\b(others|them|him|her|[A-Z][a-z]+)\b/i,
        /\b(bomb|explode|destroy|weapon|gun|knife)\b/i,
    ],
    despair: [
        /\b(no hope|hopeless|give up|can't go on|nothing matters)\b/i,
        /\b(everything is dark|drowning|trapped)\b/i,
    ],
    substance_abuse: [
        /\b(using\s+(drugs?|substances?|pills?)\b.*\b(again|relapse|overdose)?\b)/i,
        /\b(drunk\s+on\s+(substances?|drugs?)\b)/i,
        /\b(drinking\s+(alcohol|substances?)\b)/i,
        /\b(bottle\s+of\s+(pills?|medication|substances?)\b)/i,
    ],
    medical: [
        /\b(chest pain|can'?t breathe|heart attack|seizure|stroke)\b/i,
        /\b(emergency|ambulance|911|hospital)\b/i,
    ],
}

/**
 * Detects psychiatric and medical crisis signals in text
 */
export async function detectCrisisSignals(text: string): Promise<CrisisAnalysisResult> {
    // -------------------------------------------------------------------------
    // 1️⃣ PHI & Consent Safeguards
    // -------------------------------------------------------------------------
    if (!text) {
        return {
            hasCrisisSignal: false,
            riskLevel: 'minimal',
            confidence: 1.0,
            signals: [],
            actionRequired: false,
            escalationProtocol: [],
            // No approval needed for empty input
            requiresClinicianApproval: false,
            explanation: 'No text provided – minimal risk.',
        }
    }

    // Verify that the caller has consent to analyse this text.
    // In a real implementation this would check user‑level consent records.
    if (!(await ConsentChecker.hasConsentForText(text))) {
        logger.warn('Consent check failed – raw PHI not processed')
        // Return a minimal result without storing any PHI
        return {
            hasCrisisSignal: false,
            riskLevel: 'minimal',
            confidence: 1.0,
            signals: [],
            actionRequired: false,
            escalationProtocol: [],
            explainability: {
                matchedPatterns: [],
                severityFactors: [],
                matchedTexts: [],
                categories: [],
            },
        }
    }

    // Emit an audit event that PHI is being accessed (hashed for privacy)
    const auditHash = hashString(text)
    logger.info('CrisisDetectionPHIAccess', { hashedText: auditHash })

    // -------------------------------------------------------------------------
    // 2️⃣ Core Detection Logic
    // -------------------------------------------------------------------------
    const signals: CrisisSignal[] = []
    let maxSeverity = 0
    const matchedDetails: string[] = [] // collect details for explanation

    // Collect explainability data while detecting matches
    const explainabilityData: CrisisExplainability = {
        matchedPatterns: [],
        severityFactors: [],
        matchedTexts: [],
        categories: [],
    }

    for (const [category, patterns] of Object.entries(CRISIS_PATTERNS)) {
        for (const pattern of patterns) {
            const match = text.match(pattern)
            if (match) {
                const severity = calculateSeverity(category as CrisisSignal['category'], text)
                maxSeverity = Math.max(maxSeverity, severity)

                const keyword = match[0]
                matchedDetails.push(`${category} ("${keyword}")`)

                // Redact PHI in the snippet before storing
                const redactedSnippet = ProtectedHealthData.redact(
                    text.substring(
                        Math.max(0, match.index! - 30),
                        Math.min(text.length, match.index! + keyword.length + 30)
                    )
                )

                signals.push({
                    category: category as CrisisSignal['category'],
                    severity,
                    keywords: [ProtectedHealthData.redact(keyword)],
                    contextSnippet: redactedSnippet,
                })
            }
        }
    }

    const riskLevel = determineRiskLevel(maxSeverity, signals.length)
    const actionRequired = ['moderate', 'high', 'imminent'].includes(riskLevel) && isEscalationAllowed()

    const result: CrisisAnalysisResult = {
        hasCrisisSignal: signals.length > 0,
        riskLevel,
        confidence,
        signals,
        actionRequired,
        humanReviewRequired,
        escalationProtocol: generateEscalationProtocol(riskLevel, signals),
        // New fields for clinical safeguards
        requiresClinicianApproval: ['high', 'imminent'].includes(riskLevel),
        explanation: `Matched patterns: ${matchedDetails.join(', ')}`,
    }

    if (actionRequired) {
        logger.warn('Crisis signal detected', { riskLevel, signalCount: signals.length })
    }

    // Emit audit event for PHI access
    logger.audit('CrisisDetectionService.analyze', {
        riskLevel,
        snippetCount: signals.length,
        originalTextLength: text.length,
    })

    return result
}

/** Placeholder for clinician override – replace with real override mechanism. */
function isEscalationAllowed(): boolean {
    // TODO: Integrate clinician override flag (e.g., feature flag or UI confirmation)
    return true; // Default to allow escalation until override logic is implemented
}

function calculateSeverity(category: CrisisSignal['category'], text: string): number {
    // Base severity varies by category to allow a 'low' risk tier
    const CATEGORY_BASE_SEVERITY: Record<CrisisSignal['category'], number> = {
        self_harm: 0.7,
        violence: 0.6,
        despair: 0.5,
        substance_abuse: 0.3,
        medical: 0.4,
    };
    let score = CATEGORY_BASE_SEVERITY[category] ?? 0.5;

    // Specificity increases severity
    if (/\b(plan|how|where|when|tonight|today)\b/i.test(text)) score += 0.3;
    if (/\b(have|got)\b.*\b(gun|pills|knife)\b/i.test(text)) score += 0.4;

    // Immediacy
    if (/\b(now|right now|immediately)\b/i.test(text)) score += 0.2;

    return Math.min(1.0, score);
}

function determineRiskLevel(severity: number, count: number): CrisisRiskLevel {
    if (severity > 0.9) return 'imminent'
    if (severity > 0.7 || count > 2) return 'high'
    if (severity > 0.4) return 'moderate'
    if (severity > 0.2) return 'low'
    return 'minimal'
}

/**
 * Provides a confidence estimate based on the number of detected signals.
 * Capped at 0.95 to prevent over‑confidence.
 */
function calculateConfidence(signals: CrisisSignal[]): number {
    if (signals.length === 0) return 1.0;
    return Math.min(0.95, 0.7 + (signals.length * 0.1));
}

/**
 * Generates the escalation protocol for a given risk level.
 * The protocol is returned as a list of action items that must be
 * reviewed by a clinician before execution.
 */
function generateEscalationProtocol(riskLevel: CrisisRiskLevel, _signals: CrisisSignal[]): string[] {
    const protocol: string[] = []

    if (riskLevel === 'imminent') {
        protocol.push('Contact emergency services (911/112) immediately')
        protocol.push('Notify on-call psychiatrist')
        protocol.push('Activate immediate location tracking')
    } else if (riskLevel === 'high') {
        protocol.push('Contact primary therapist for urgent intervention')
        protocol.push('Notify crisis response team')
        protocol.push('Initiate safety plan review')
    } else if (riskLevel === 'moderate') {
        protocol.push('Flag session for supervisor review')
        protocol.push('Schedule follow-up within 24 hours')
        protocol.push('Provide crisis resources to user')
    }

    return protocol
}